### PR TITLE
Fix oversized items not fitting into a chunk

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

There is an edge case in DataProcessor/Cache when the user returns an item that is larger (in bytes) than the selected chunk size. Previously, this would cryptically error at runtime (hours in) and the user would have to restart with a larger chunk size configured. Now, we avoid this error by making the chunk bigger to fit the element, but we also emit a warning. 

----
:books: Documentation preview :books:: https://pytorch-lightning--18938.org.readthedocs.build/en/18938/

### CODE REVIEW

File: src/lightning/data/streaming/writer.py [hunk 1]
```diff
@@ -212,39 +210,42 @@ def _serialize(self, item: Any, sizes: List[int], data: List[bytes]) -> str:
...
         """Create a binary chunk from all the binarized items."""
+        items = []
+
         if on_done:
             indices = sorted(self._serialized_items.keys())
             for i in range(len(indices) - 1):
                 assert indices[i] == indices[i + 1] - 1, indices
-            min_index = indices[0]
-            max_index = indices[-1] + 1
-            num_items = np.uint32(max_index - min_index)
```
Comment (awaelchli) [inline src/lightning/data/streaming/writer.py:221]: Correct me if I'm wrong, but num_items is always len(items) right? So we don't need to compute it this complicated way, and I just moved the definition below.

File: src/lightning/data/streaming/writer.py [hunk 2]
```diff
@@ -212,39 +210,42 @@ def _serialize(self, item: Any, sizes: List[int], data: List[bytes]) -> str:
...
                 f" Found {self._pretty_serialized_items()} with boundaries: {self._min_index}, {self._max_index}."
             )
 
+        num_items = np.uint32(len(items))
         sizes = list(map(len, items))
         offsets = np.array([0] + sizes).cumsum().astype(np.uint32)
         offsets += len(num_items.tobytes()) + len(offsets.tobytes())
         sample_data = b"".join([item.data for item in items])
         data = num_items.tobytes() + offsets.tobytes() + sample_data
-        offsets = offsets.tolist()
```
Comment (awaelchli) [inline src/lightning/data/streaming/writer.py:242]: this was unused

File: tests/tests_data/streaming/test_cache.py [hunk 3]
```diff
@@ -242,3 +243,31 @@ def test_streaming_dataset(tmpdir, monkeypatch):
...
+        cache[3] = np.random.randint(0, 10, size=(200,), dtype=np.uint8)
+
+    cache.done()
+    cache.merge()
+    assert len(os.listdir(tmp_path)) == 4  # 3 chunks + 1 index file
+    metadata_bytes = 167
+    assert os.path.getsize(tmp_path / "chunk-0-0.bin") == 500 + metadata_bytes
+    assert os.path.getsize(tmp_path / "chunk-0-1.bin") == 10000 + metadata_bytes
+    # TODO: Why doesn't this assertion hold?
+    # assert os.path.getsize(tmp_path / "chunk-0-2.bin") == 150 + 200 + metadata_bytes
```
Comment (awaelchli) [inline tests/tests_data/streaming/test_cache.py:273]: @tchaton I have no idea why the file size isn't right here

File: src/lightning/data/streaming/writer.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/data/streaming/writer.py:221]: Yes.

File: tests/tests_data/streaming/test_cache.py (hunk 3 above)
Comment (tchaton) [reply tests/tests_data/streaming/test_cache.py:273]: What's the size ?

Comment (awaelchli) [reply tests/tests_data/streaming/test_cache.py:273]: 668
if we subtract these 167 bytes that the others have extra, we get 501 🤔

Comment (tchaton) [reply tests/tests_data/streaming/test_cache.py:273]: I think it is the pickle overhead. We don't have a numpy serializer.

Comment (awaelchli) [reply tests/tests_data/streaming/test_cache.py]: I revised the test, asserting the number of items per chunk via the config file

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by rusenask — APPROVED** [review]

**Pull request merged**